### PR TITLE
Add Tor client to Stack Wallet

### PR DESCRIPTION
**Pull request by sneurlax**
This PR adds an experimental Tor client to Stack Wallet

It uses Arti, the Tor Project's new client written in Rust

It is also compatible with the old C-based Tor if backwards compatibility is needed

TODO:
- [x] Mobile UI
- [x] Support exchange .onion services
- [x] Check build setup for macOS
- [x] Check build setup for iOS
- [x] Check build setup for Android
- [x] Check build setup for Windows

BONUS:
- [ ] Tor circuit / validity check

Screenshot from 2023-09-07 18-25-42
Screenshot from 2023-09-07 18-28-01
Screenshot from 2023-09-07 18-27-51

Screenshot from 2023-09-08 16-27-09
Screenshot from 2023-09-08 16-27-13
Screenshot from 2023-09-08 16-27-32

### CODE REVIEW

**Comment by sneurlax** [comment]
Closing this to stop running tasks, but I'll edit the top post with tasks for merging this in.

File: .gitignore [hunk 1]
```diff
@@ -55,4 +55,10 @@ libcw_monero.dll
 libcw_wownero.dll
 libepic_cash_wallet.dll
 libmobileliblelantus.dll
+libtor_ffi.dll
 /libisar.so
+libtor_ffi.so
+
+tor_logs.txt
```
Comment (sneurlax) [inline .gitignore:62]: These last two lines added to .gitignore could be left off

File: .gitmodules
Comment (sneurlax) [inline .gitmodules:1]: What even is this change?

File: lib/electrumx_rpc/cached_electrumx.dart
Comment (sneurlax) [inline lib/electrumx_rpc/cached_electrumx.dart:1]: I forget why these firo-related changes aren't in staging already

File: lib/networking/socks_socket.dart
Comment (sneurlax) [inline lib/networking/socks_socket.dart:1]: This could be spun out into its own package.

File: lib/networking/socks_socket.dart [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+      },
+      onError: (e) {
+        // Handle errors.
+        if (e is Object) {
+          _responseController.addError(e);
+        }
+
+        // If the error is not an object, send the error as a string.
+        _responseController.addError("$e");
+        // TODO make sure sending error as string is acceptable.
```
Comment (sneurlax) [inline lib/networking/socks_socket.dart:158]: How would we test if Stack (or Stack's tor?) can or can't handle strings?

File: lib/networking/socks_socket.dart [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+        _responseController.close();
+      },
+    );
+  }
+
+  /// Connects to the SOCKS socket.
+  ///
+  /// Returns:
+  ///  A Future that resolves to void.
+  Future<void> connect() async {
```
Comment (sneurlax) [inline lib/networking/socks_socket.dart:171]: connect doesn't return.

File: lib/networking/socks_socket.dart [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+          // If the error is not an object, send the error as a string.
+          _secureResponseController.addError("$e");
+          // TODO make sure sending error as string is acceptable.
+        },
+        onDone: () {
+          // Close the response controller when the socket is closed.
+          _secureResponseController.close();
+        },
+      );
+    }
```
Comment (sneurlax) [inline lib/networking/socks_socket.dart:248]: OK, if `sslEnabled` then `_subscription = _secureSocksSocket.listen(`, but what about non-SSL connections?

We still need to `_subscription = _socksSocket.listen(` for non-SSL connections, right?

File: lib/networking/socks_socket.dart [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+
+  /// Sends the server.features command to the proxy server.
+  ///
+  /// This method is used to send the server.features command to the proxy
+  /// server. This command is used to request the features of the proxy server.
+  /// It serves as a demonstration of how to send commands to the proxy server.
+  ///
+  /// Returns:
+  ///   A Future that resolves to void.
+  Future<void> sendServerFeaturesCommand() async {
```
Comment (sneurlax) [inline lib/networking/socks_socket.dart:282]: This should probably be moved to the bottom of the class.  It isn't used except by an example in another repository...  It should've been moved into the tor package and used through that

File: lib/networking/socks_socket.dart
Comment (sneurlax) [inline lib/networking/socks_socket.dart:1]: This should be spun out into its own package and/or put into and used through the tor / tor_ffi_plugin package.

File: lib/pages/exchange_view/exchange_coin_selection/exchange_currency_selection_view.dart [hunk 6]
```diff
@@ -39,6 +40,8 @@ import 'package:stackwallet/widgets/stack_dialog.dart';
 import 'package:stackwallet/widgets/stack_text_field.dart';
 import 'package:stackwallet/widgets/textfield_icon_button.dart';
 
+import '../../../services/exchange/exchange.dart';
```
Comment (sneurlax) [inline lib/pages/exchange_view/exchange_coin_selection/exchange_currency_selection_view.dart:43]: This should use a package import

File: lib/pages/ordinals/ordinal_details_view.dart [hunk 7]
```diff
@@ -229,12 +231,18 @@ class _OrdinalImageGroup extends StatelessWidget {
...
-          "statusCode=${response.statusCode} body=${response.bodyBytes}");
+    final response = await client.get(
+      url: Uri.parse(ordinal.content),
+      proxyInfo: ref.read(prefsChangeNotifierProvider).useTor
+          ? ref.read(pTorService).getProxyInfo()
+          : null,
+    );
+
+    if (response.code != 200) {
+      throw Exception("statusCode=${response.code} body=${response.bodyBytes}");
```
Comment (sneurlax) [inline lib/pages/ordinals/ordinal_details_view.dart:245]: `throw Exception("OrdinalDetailsView _savePngToFile statusCode=${response.code} body=${response.bodyBytes}");`

File: lib/pages/settings_views/wallet_settings_view/wallet_settings_view.dart
Comment (sneurlax) [inline lib/pages/settings_views/wallet_settings_view/wallet_settings_view.dart:1]: Do we intend to remove firo cache-clearing here?

File: lib/pages/wallet_view/transaction_views/transaction_details_view.dart [hunk 8]
```diff
@@ -358,6 +358,8 @@ class _TransactionDetailsViewState
     final currentHeight = ref.watch(walletsChangeNotifierProvider
         .select((value) => value.getManager(walletId).currentHeight));
 
+    print("THIS TRANSACTION IS $_transaction");
```
Comment (sneurlax) [inline lib/pages/wallet_view/transaction_views/transaction_details_view.dart:361]: no prints in prod :)

File: pubspec.lock
Comment (sneurlax) [inline pubspec.lock:1]: Should we limit these pubspec.lock changes?  I `flutter upgrade`d to refresh my tor_ffi_plugin (cypherstack/tor) package (which I updated to export SOCKSSocket and return from connect)

**Review by julian-CStack — CHANGES_REQUESTED** [review]

File: lib/pages/wallet_view/sub_widgets/tx_icon.dart
[the remark this answers is not in the document]
Comment (sneurlax) [reply lib/pages/wallet_view/sub_widgets/tx_icon.dart:1]: Resolved by https://github.com/cypherstack/stack_wallet/pull/664/commits/8d6c355b9598f360a6c623d062405d1e5f9e9b2b

File: windows/flutter/generated_plugins.cmake
[the remark this answers is not in the document]
Comment (sneurlax) [reply windows/flutter/generated_plugins.cmake:1]: Yes, and these additional changes are also needed: https://github.com/cypherstack/stack_wallet/pull/664/commits/cd00e9f8fd6bfe085dbe85d920ca6eca61230751

**Pull request merged**